### PR TITLE
[MAD][DATA1018][lab-sklearn-and-unsupervised-learning] Maris Font

### PR DESCRIPTION
**Pull request by marisfont**
Hola Belen - terminado, gracias por todo

### CODE REVIEW

**Comment by Gobinde43** [comment]
> Hola Belen - terminado, gracias por todo

It's been a pleasure!! Good luck Maris and keep in touch!!

File: module-3/lab-sklearn-and-unsupervised-learning/your-code/main.ipynb [hunk 1]
```diff
@@ -2837,11 +2837,37 @@
   },
```
Comment (Gobinde43) [inline module-3/lab-sklearn-and-unsupervised-learning/your-code/main.ipynb]: - "What is the average ash content for each cluster?". Another way of doing it:
```Python
(np.average(wine['data'][:,2][np.where(kmeans.labels_ == 0)]), np.average(wine['data'][:,2][np.where(kmeans.labels_ == 1)]), 
np.average(wine['data'][:,2][np.where(kmeans.labels_ == 2)]), np.average(wine['data'][:,2][np.where(kmeans.labels_ == 3)]))
```
- "For the prescribed_medicines column, fill the missing values with the value no prescription. For the diagnosis column, fill the missing values with no diagnosis." better in one line when possible and the code is readable enough:
```Python
patients.fillna({'prescribed_medicines': 'no prescription', 'diagnosis': 'no diagnosis'}, inplace=True)
```
- Good to use the label encoder. Other way of doing it without a for loop:
```Python
#your code 
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
for column in patients.columns:
    if patients[column].dtype == bool: 
        le.fit(patients[column])
        patients[column] = le.transform(patients[column])
#Other way 
from sklearn.preprocessing import LabelEncoder
label = LabelEncoder()
label.fit([False, True])
patients[['patient_diabetic', 'patient_allergic', 'patient_show', 'is_regular_visit']] = patients[['patient_diabetic', 'patient_allergic', 'patient_show', 'is_regular_visit']].apply(label.fit_transform)
patients.head()
```
- In this one "The last step is to handle the object data." You can simplify your code in the following way:
```Python
#Your code 
for column in patients.columns:
    if patients[column].dtype == object: 
        print(patients[column].value_counts())
# Other way 
print(patients.patient_gender.value_counts())

print('-----')

print(patients.doctor_name.value_counts())

print('-----')

print(patients.prescribed_medicines)

print('-----')

print(patients.diagnosis.value_counts())
```
- "Bonus Challenge: Visualize K-Means Clusters" , another way of doing this plot:
```Python
fig = plt.figure()

ax = fig.add_subplot(111)

scatter = ax.scatter(patients['patient_weight_kg'], patients['patient_height_sm'],
                     c=patients['labels'])

plt.colorbar(scatter)
```
- Great job in the 3D plot. In general terms, the only problem with 3D plots is that human brain doesn't work properly in 3D.
- We can divide (as a taxonomy) the clustering algorithms in two classes:
1. Clustering based in distance: observations can be represented as points in a dimensional space where the dimensions are the features we have. The distance metric can be (euclidean, manhattan, mahalanobis) to separate the different clusters. K-means belongs to this class and it's the most famous one.
2. Clustering based on density: this algorithms don't into account only the distance among "points" but also detecting the areas where the density of points is higher.
- Find below a link to the main clustering algorithms:
https://towardsdatascience.com/the-5-clustering-algorithms-data-scientists-need-to-know-a36d136ef68
- Sklearn has all this clustering algorithms:
https://scikit-learn.org/stable/modules/classes.html#module-sklearn.cluster
